### ISSUE TITLE
Illogical search results (to me!)

### ISSUE THREAD

**Issue by trailerparkdev** [meilisearch/meilisearch#2378]
Hi Meilisearch, and thanks for the good work.

Could you explain to me why in searching "personnel reprographie" (with CURL, against v0.26), I get

```
{
  "hits": [
    {
      "lemmatized": "convention collectif national commerce détail papeterie fourniture bureau bureautique informatique librairie 15 décembre 1988 étendre arrêter 14 décembre 1989 jorf 30 décembre 1989 champ application fusionner celui convention collectif national personnel reprographie idcc 706 arrêter ministériel 9 avril 2019",
      "uid": "ebd824693c42d7b9351d934aa0dc000caa7ae6ed53b93472f9c5918ac4ed9d5f"
    },
    {
      "lemmatized": "convention collectif national personnel société coopératif hlm 15 mai 1990 champ application fusionner celui convention collectif national personnel office public habitat désormais intituler convention collectif national personnel office public habitat société coordination idcc 3220 arrêter ministériel 16 novembre 2018",
      "uid": "ba7f340219d108c8d39071cfd883e8922d930f2a964241f872bbf16b51a05953"
    }
  ],
  "nbHits": 2,
  "exhaustiveNbHits": false,
  "query": "personnel reprographie",
  "limit": 20,
  "offset": 0,
  "processingTimeMs": 4
}

```

and when I search "accord personnel reprographie", I get no hits ?

This seems strange to me: 

Result #2 has only 1 word matching ("personnel") and nothing close to the second word ("reprographie"), but I get the result anyway. So far so good.

So why with adding "accord" to the query don't I get at least result #1, which has both "personnel" and "reprographie" ?  

Before bothering you, I have tried with all combinations of ranking rules, including default: no love. 

Thank you for your time answering this.

Have a nice day.

**Comment by curquiza** [comment]
Hello @trailerparkdev

Thanks for trying Meilisearch and welcome to the community 😄 

To retrieve your documents, Meilisearch works in a specific way, performing
- the prefix search: we supposed every word typed in the query is a word you want to retrieve in the documents, knowing the latest typed word (the one on the right for you) could have been partially typed
- the words ranking rule: the latest words typed should not be necessary in the documents, the first typed words are the most important.

What does it mean for your usecase?
Following your query `accord personnel reprographie`, Meilisearch will retrieve all your documents containing `accord` AND `personnel` AND `reprographie`. Then, if there are not enough results, it will retrieve all your documents containing `accord` AND `personnel`. And finally, if there are still not enough results, it will retrieve all your documents containing only `accord`.
As you can see, since you put `accord` as the first word of your query, Meilisearch will always try to get documents containing at least `accord`.
If you type this query: `personnel reprographie accord ` instead of `accord personnel reprographie`, both of your documents will be present in the results.

More information about
- [prefix search](https://docs.meilisearch.com/learn/advanced/prefix.html#prefix-search)
- [words ranking rules](https://docs.meilisearch.com/learn/core_concepts/relevancy.html#_1-words)
Hope it helps!

**Comment by trailerparkdev** [comment]
Thank you @curquiza for your quick and detailed reply.

I understand the rationale and I should have remembered that MeiliSearch implements a prefix search (it's in the docs) .

Would you allow me to share a more global point of view?

This prefix search example -- and many others in milli / MeiliSearch -- indicate an opinionated platform. 

As developers, we all know that opinionated is good because it simplifies things and reduces time to deployment, as long as the use case fits. 

Now, implementing opinions as default BUT allowing non-defaults (in this case regular vs prefix search) would benefit both you and us: we would have more flexibility to adapt to usecase specificities (ie the real life! [1]) and you would make your platform more universal (ie somehow lift the "if usecase fits" barrier). 

In other words, defaults make MeiliSearch a very valuable and efficient open-source Algolia with a great future. Non-defaults alternatives would open MeiliSearch to a lot more usecases based on the likes of ElasticSearch (that many want to dump for more efficiency). 

I don't want to sound sound like I'm telling you what to do. But I believe allowing non-defaults is (generally, based on your clean code) a small effort with a great return on time investment. Opinionated is good. Flexibly opinionated is the best of both worlds.

Have a nice day.

[1] Go explain customers and their users that they must enter the most important search tokens first...

**Comment by Kerollmops** [comment]
Hey @trailerparkdev,

For your information Meilisearch is an open source project and we are open to any feature that would help people using the engine.

Therefore, this is why you can propose any feature you like with an explanation about your use case in [this dedicated repository](https://github.com/meilisearch/product/discussions).

There is no reason to explain why what we have done is wrong and that we should implement something or something more, just propose that on the appropriate repository and wait for our feedback. We are doing a lot of work in specifying the settings and default options the engine should have, don't say the contrary.

Have a nice day!

**Comment by trailerparkdev** [comment]
Hi @Kerollmops.

Thanx for the time of your reply.

Duly noted for the repository.

I didn't mean to imply that what you've done so far is "wrong" in any way. If that's what it sounded like, I'm sorry. And yes there are many instances of your specifying settings and default options, and you do that with a remarkable documentation (concise and precise). I don't have the impression of having stated the contrary. 

My only point was that, from a peer developer and friendly standpoint, cornering MS to mainly search-as-you-type use cases seemed limitative considering its capabilities and the talent of the team (from what the code reveals). In this respect, allowing for non-prefix searches would immediately open MS to many new use cases, with what I believe is not a inhuman effort, even in a startup's constrained time environment.

These were just my two cents, initially proposed to you guys with nothing but a positive attitude. 

Keep up the good work.

**Comment by Kerollmops** [comment]
Hey @trailerparkdev,

No worries, we will review your feature request, post it in the product repository when you have time 🙂

> In this respect, allowing for non-prefix searches would immediately open MS to many new use cases

Have you tried to [enclose the query words in double-quotes](https://docs.meilisearch.com/reference/api/search.html#search-in-an-index-with-get-route) already? I know it can be cumbersome but it is at least a good enough solution for now.

Have a good day!

**Comment by trailerparkdev** [comment]
Hi @Kerollmops.

Thanks again for responding. 

The double-quotes trick may not always give expected results with search patterns whose tokens do not really follow one another in the target documents.

Will post the FR!

Have a nice night.

**Comment by curquiza** [comment]
Thank you @trailerparkdev for posting your feature request!

I close this issue, feel free to re-open if needed!
